src/main/java/com/google/cloud/spanner/jdbc/JdbcTypeConverter.java

```diff
@@ -163,9 +170,61 @@ static Object convert(Object value, Type type, Class<?> targetType) throws SQLEx
               ZoneId.systemDefault());
         }
       }
+      if (AbstractMessage.class.isAssignableFrom(targetType)) {
+        if (type.getCode() == Code.PROTO || type.getCode() == Code.BYTES) {
+          Method parseMethodParseFrom = targetType.getMethod("parseFrom", byte[].class);
+          return targetType.cast(parseMethodParseFrom.invoke(null, value));
+        }
+      }
+      if (ProtocolMessageEnum.class.isAssignableFrom(targetType)) {
+        if (type.getCode() == Code.ENUM || type.getCode() == Code.INT64) {
+          Method parseMethodForNumber = targetType.getMethod("forNumber", int.class);
+          return targetType.cast(
+              parseMethodForNumber.invoke(
+                  null, AbstractJdbcWrapper.checkedCastToInt((Long) value)));
+        }
+      }
       if (targetType.equals(java.sql.Array.class)) {
         if (type.getCode() == Code.ARRAY) return value;
       }
+      if (targetType.isArray() && type.getCode() == Code.ARRAY) {
+        if (type.getArrayElementType().getCode() == Code.PROTO
+            || type.getArrayElementType().getCode() == Code.BYTES) {
+          Class<?> componentType = targetType.getComponentType();
+          if (AbstractMessage.class.isAssignableFrom(componentType)) {
+            byte[][] result = (byte[][]) ((JdbcArray) value).getArray();
+            Object obj = java.lang.reflect.Array.newInstance(componentType, result.length);
+            for (int i = 0; i < result.length; i++) {
+              if (result[i] != null) {
+                Method parseMethodParseFrom = componentType.getMethod("parseFrom", byte[].class);
+                java.lang.reflect.Array.set(
+                    obj, i, componentType.cast(parseMethodParseFrom.invoke(null, result[i])));
+              }
+            }
+            return obj;
+          }
+        }
+        if (type.getArrayElementType().getCode() == Code.ENUM
+            || type.getArrayElementType().getCode() == Code.INT64) {
+          Class<?> componentType = targetType.getComponentType();
+          if (ProtocolMessageEnum.class.isAssignableFrom(componentType)) {
+            Long[] result = (Long[]) ((JdbcArray) value).getArray();
+            Object obj = java.lang.reflect.Array.newInstance(componentType, result.length);
+            for (int i = 0; i < result.length; i++) {
+              if (result[i] != null) {
+                Method parseMethodForNumber = componentType.getMethod("forNumber", int.class);
+                java.lang.reflect.Array.set(
+                    obj,
+                    i,
+                    componentType.cast(
+                        parseMethodForNumber.invoke(
+                            null, AbstractJdbcWrapper.checkedCastToInt(result[i]))));
+              }
+            }
+            return obj;
+          }
+        }
+      }
```

Comment:
Could these two blocks be put in separate methods? They are significantly larger/more complex than most of the other blocks in this method, so deviating from that would make sense.

Reply:
Done